src/lib/Algo.hs

```diff
@@ -287,17 +290,20 @@ select (ps, _) (qs, _) = compareCons (head ps) (head qs)
 -- | Tests whether the given patterns match the same constructor or both match
 --   the wildcard pattern.
 compareCons :: Pat () -> Pat () -> Bool
-compareCons (PApp _ qn1 _       ) (PApp _ qn2 _       ) = qn1 == qn2
-compareCons (PInfixApp _ _ qn1 _) (PInfixApp _ _ qn2 _) = qn1 == qn2
-compareCons (PApp _ qn1 _       ) (PInfixApp _ _ qn2 _) = qn1 == qn2
-compareCons (PInfixApp _ _ qn1 _) (PApp _ qn2 _       ) = qn1 == qn2
-compareCons (PParen _ p1        ) p2                    = compareCons p1 p2
-compareCons p1                    (PParen _ p2   )      = compareCons p1 p2
--- TODO Special Syntax
-compareCons (PList _ ps1) (PList _ ps2) = length ps1 == length ps2
-compareCons (PWildCard _   )      (PWildCard _   )      = True
-compareCons (PTuple _ _ ps1) (PTuple _ _ ps2) = length ps1 == length ps2
-compareCons _                     _                     = False
+compareCons = (==) `on` consName
+
+-- | Returns the qualified name of the constructor of the given pattern or
+--   @Nothing@ if it is a wildcard pattern.
+consName :: Pat () -> Maybe (QName ())
+consName (PApp _ qn _       ) = return qn
+consName (PInfixApp _ _ qn _) = return qn
+consName (PParen _ pat      ) = consName pat
+consName (PList  _ []       ) = return $ Special () $ ListCon ()
+consName (PList  _ (_ : _)  ) = return $ Special () $ Cons ()
+consName (PTuple _ _ ps) = return $ Special () $ TupleCon () Boxed $ length ps
+consName (PWildCard _       ) = Nothing
+consName pat =
+  error $ "consName: unsupported pattern \"" ++ prettyPrint pat ++ "\""
```

Comment:
@macmaccamcam do you think it would be possible to reuse this function in [`getQNamePat`](https://github.com/FreeProving/haskell-src-transformations/blob/33c2fe955539f7be6422f5fb6629234aaf8ec1c2/src/lib/Algo.hs#L205)?

Reply:
I can't think of any reason why it shouldn't.